esmvalcore/preprocessor/_multimodel.py

```diff
@@ -201,158 +185,102 @@ def _put_in_cube(template_cube, cube_data, statistic, t_axis):
-            ndat.mask[indices[idx_cube]] = cube.data.mask
-        else:
-            ndat.mask[indices[idx_cube]] = False
-        ndatarr[idx_cube] = ndat[t_idx]
-
-    # return time slice
-    return ndatarr
-
+    t_unit = _get_consistent_time_unit(cubes)
 
-def _assemble_overlap_data(cubes, interval, statistic):
-    """Get statistical data in iris cubes for OVERLAP."""
-    start, stop = interval
-    sl_1, sl_2 = _slice_cube(cubes[0], start, stop)
-    stats_dats = np.ma.zeros(cubes[0].data[sl_1:sl_2 + 1].shape)
-
-    # keep this outside the following loop
-    # this speeds up the code by a factor of 15
-    indices = [_slice_cube(cube, start, stop) for cube in cubes]
-
-    for i in range(stats_dats.shape[0]):
-        time_data = [
-            cube.data[indx[0]:indx[1] + 1][i]
-            for cube, indx in zip(cubes, indices)
-        ]
-        stats_dats[i] = _compute_statistic(time_data, statistic)
-    stats_cube = _put_in_cube(cubes[0][sl_1:sl_2 + 1],
-                              stats_dats,
-                              statistic,
-                              t_axis=None)
-    return stats_cube
-
-
-def _assemble_full_data(cubes, statistic):
-    """Get statistical data in iris cubes for FULL."""
-    # all times, new MONTHLY data time axis
-    time_axis = [float(fl) for fl in _monthly_t(cubes)]
+    for cube in cubes:
+        # Extract date info from cube
+        years = [cell.point.year for cell in cube.coord('time').cells()]
+        months = [cell.point.month for cell in cube.coord('time').cells()]
+
+        # Reconstruct default calendar
+        if 0 not in np.diff(years):
+            # yearly data
+            dates = [datetime(year, 7, 1) for year in years]
+
+        elif 0 not in np.diff(months):
+            # monthly data
+            dates = [
+                datetime(year, month, 15)
+                for year, month in zip(years, months)
+            ]
+        else:
+            # (sub)daily data
+            if cube.coord('time').units != t_unit:
+                logger.warning(
+                    "Multimodel encountered (sub)daily data and inconsistent "
+                    "time units or calendars. Attempting to continue, but "
+                    "might produce unexpected results.")
+            dates = [cell.point for cell in cube.coord('time').cells()]
+
+        # Update the cubes' time coordinate (both point values and the units!)
+        cube.coord('time').points = [t_unit.date2num(date) for date in dates]
+        cube.coord('time').units = t_unit
+        cube.coord('time').bounds = None
+        cube.coord('time').guess_bounds()
+
+
+def _get_time_slice(cubes, time):
+    """Fill time slice array with cubes' data if time in cube, else mask."""
+    time_slice = []
+    for cube in cubes:
+        cube_time = cube.coord('time').points
+        if time in cube_time:
+            idx = int(np.argwhere(cube_time == time))
+            subset = cube[idx].data
```

Comment:
```suggestion
            subset = cube.data[idx]
```
would be faster, because slicing a numpy array is less work than slicing an entire cube and you're only using the numpy array anyway